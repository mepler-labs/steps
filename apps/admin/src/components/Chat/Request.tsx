import DateDisplay from 'atoms/DateDisplay';
import Panel from 'atoms/Panel';
import Status from 'atoms/Status';
import { Flex } from 'grid-styled';
import moment from 'moment';
import React from 'react';
import styled from 'styled-components';
import { blue, green, red } from 'styles/colors';
import { RequestStatus } from './types';

interface Props {
  className?: string;
  status: RequestStatus;
  message: string;
  date: moment.Moment;
}

const statusDetails = {
  NEEDS_ASSISTANCE: {
    text: 'Needs assistance',
    color: red,
  },
  REPLIED: {
    text: 'Replied',
    color: blue,
  },
  RESOLVED: {
    text: 'Resolved',
    color: green,
  },
};

const statusText = status => statusDetails[status].text;
const statusColor = status => statusDetails[status].color;

export const Request: React.SFC<Props> = ({
  className,
  status,
  message,
  date,
}) => (
  <StyledRequest className={className} status={status}>
    <Panel className="container">
      <Flex alignItems="center" justifyContent="space-between">
        <Status color={statusColor(status)}>{statusText(status)}</Status>
        <DateDisplay date={date} />
      </Flex>
      <Message>{message}</Message>
    </Panel>
  </StyledRequest>
);

export const Message = styled.div`
  font-size: 30px;
`;

const StyledRequest = styled<{ status: string }, 'div'>('div')`
  border-left: 10px solid ${({ status }) => statusColor(status)};
  font-family: 'Calibre', sans-serif;
  margin: 20px;

  .container {
    padding: 30px;
  }
`;

export default Request;
