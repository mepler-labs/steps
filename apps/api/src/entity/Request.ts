import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToOne,
  JoinColumn,
  OneToMany
} from "typeorm";
import { Message } from "./Message";

// also enforced by CHECK constraint
export type RequestStatus = "NEEDS_ASSISTANCE" | "REPLIED" | "RESOLVED";

@Entity("request")
export class Request {
  @PrimaryGeneratedColumn() id: number;

  @Column() status: RequestStatus;

  @OneToMany(type => Message, message => message.request)
  messages: Message[];
}
