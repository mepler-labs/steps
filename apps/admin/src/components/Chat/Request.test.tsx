import { shallow } from 'enzyme';
import { Request } from './Request';
import * as React from 'react';
import 'jest';
import moment from 'moment';

const date = moment.utc(new Date(2018, 5, 7));

describe('Request', () => {
  it('shows request status', () => {
    const wrapper = shallow(
      <Request status="NEEDS_ASSISTANCE" message="I need help" date={date} />,
    );

    expect(wrapper.find('.status').text()).toBe('Needs assistance');
  });

  it('shows message', () => {
    const wrapper = shallow(
      <Request status="NEEDS_ASSISTANCE" message="I need help" date={date} />,
    );

    expect(wrapper.find('.message').text()).toBe('I need help');
  });

  it('formats date', () => {
    const wrapper = shallow(
      <Request status="NEEDS_ASSISTANCE" message="I need help" date={date} />,
    );

    expect(wrapper.find('.date').text()).toBe(date.format('YYYY-M-D'));
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <Request status="NEEDS_ASSISTANCE" message="I need help" date={date} />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
