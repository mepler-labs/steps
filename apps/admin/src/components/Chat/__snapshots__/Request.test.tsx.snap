// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request renders correctly 1`] = `
<styled.div
  status="NEEDS_ASSISTANCE"
>
  <Panel
    className="container"
  >
    <Flex
      alignItems="center"
      justifyContent="space-between"
      theme={
        Object {
          "breakpoints": Array [
            "40em",
            "52em",
            "64em",
          ],
          "fontSizes": Array [
            12,
            14,
            16,
            20,
            24,
            32,
            48,
            64,
            96,
            128,
          ],
          "space": Array [
            0,
            4,
            8,
            16,
            32,
            64,
            128,
            256,
            512,
          ],
        }
      }
    >
      <Status
        color="#F73833"
      >
        Needs assistance
      </Status>
      <DateDisplay
        date={"2018-06-07T05:00:00.000Z"}
      />
    </Flex>
    <styled.div>
      I need help
    </styled.div>
  </Panel>
</styled.div>
`;
